cache/remotecache/s3/s3.go

```diff
@@ -0,0 +1,541 @@
+	AccessKeyID      string
+	SecretAccessKey  string
+	S3ForcePathStyle bool
+}
+
+func getConfig(attrs map[string]string) (*Config, error) {
+	bucket, ok := attrs[attrBucket]
+	if !ok {
+		bucket, ok = os.LookupEnv("AWS_BUCKET")
+		if !ok {
+			errors.Errorf("bucket ($AWS_BUCKET) not set for s3 cache")
+		}
+	}
+
+	region, ok := attrs[attrRegion]
+	if !ok {
+		region, ok = os.LookupEnv("AWS_REGION")
+		if !ok {
+			errors.Errorf("region ($AWS_REGION) not set for s3 cache")
+		}
+	}
+
+	role, ok := attrs[attrRole]
+	if !ok {
+		role = os.Getenv("AWS_ROLE_ARN")
+	}
+
+	prefix, ok := attrs[attrPrefix]
+	if !ok {
+		prefix = ""
+	}
+
+	manifestsPrefix, ok := attrs[attrManifestsPrefix]
+	if !ok {
+		manifestsPrefix = "manifests/"
+	}
+
+	blobsPrefix, ok := attrs[attrBlobsPrefix]
+	if !ok {
+		blobsPrefix = "blobs/"
+	}
+
+	names := []string{"buildkit"}
+	name, ok := attrs[attrName]
+	if ok {
+		splittedNames := strings.Split(name, ";")
+		if len(splittedNames) > 0 {
+			names = splittedNames
+		}
+	}
+
+	touchRefresh, _ := time.ParseDuration("24h")
+
+	touchRefreshStr, ok := attrs[attrTouchRefresh]
+	if ok {
+		touchRefreshFromUser, err := time.ParseDuration(touchRefreshStr)
+		if err == nil {
+			touchRefresh = touchRefreshFromUser
+		}
+	}
+
+	endpointURL, ok := attrs[attrEndpointURL]
+	if !ok {
+		endpointURL = ""
+	}
+
+	accessKeyID, ok := attrs[attrAccessKeyID]
+	if !ok {
+		accessKeyID = ""
+	}
+
+	secretAccessKey, ok := attrs[attrSecretAccessKey]
+	if !ok {
+		secretAccessKey = ""
+	}
+
+	s3ForcePathStyle := false
+	s3ForcePathStyleStr, ok := attrs[attrS3ForcePathStyle]
+	if ok {
+		s3ForcePathStyleUser, err := strconv.ParseBool(s3ForcePathStyleStr)
+		if err == nil {
+			s3ForcePathStyle = s3ForcePathStyleUser
+		}
+	}
+
+	return &Config{
+		Bucket:           bucket,
+		Region:           region,
+		Role:             role,
+		Prefix:           prefix,
+		ManifestsPrefix:  manifestsPrefix,
+		BlobsPrefix:      blobsPrefix,
+		Names:            names,
+		TouchRefresh:     touchRefresh,
+		EndpointURL:      endpointURL,
+		AccessKeyID:      accessKeyID,
+		SecretAccessKey:  secretAccessKey,
+		S3ForcePathStyle: s3ForcePathStyle,
+	}, nil
+}
+
+// ResolveCacheExporterFunc for s3 cache exporter.
+func ResolveCacheExporterFunc() remotecache.ResolveCacheExporterFunc {
+	return func(ctx context.Context, g session.Group, attrs map[string]string) (remotecache.Exporter, error) {
+		config, err := getConfig(attrs)
+		if err != nil {
+			return nil, err
+		}
+
+		return NewExporter(config)
+	}
+}
+
+type exporter struct {
+	solver.CacheExporterTarget
+	chains *v1.CacheChains
+	cache  *cache
+	config *Config
+}
+
+func NewExporter(config *Config) (remotecache.Exporter, error) {
+	cc := v1.NewCacheChains()
+	cache, err := newCache(config)
+	if err != nil {
+		return nil, err
+	}
+	return &exporter{CacheExporterTarget: cc, chains: cc, cache: cache, config: config}, nil
+}
+
+func (e *exporter) Config() remotecache.Config {
+	return remotecache.Config{}
```

Comment:
This should return `Compression: default`

Reply:
done